Generate trusted_root.json in the TUF server

bkabrda opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary

This PR replaces #1191 and is the last PR to address #1182 to make the TUF server useful for prod deployment - it allows for generating the `trusted_root.json` file and serving it as one of the TUF targets. It does the same as #1191, but uses the [new functionality from sigstore-go](https://github.com/sigstore/sigstore-go/commit/2198ac32dd94ce70cb0c9ee4c06ca8b5906099a9).

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

The TUF server was improved to generate a `trusted_root.json` target from the supplied files.

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->
I don't believe that the TUF server is even documented anywhere, so I don't think documentation change is required, but please let me know if I'm wrong about that.

haydentherapper commented on cmd/tuf/server/main.go:
```diff
@@ -50,8 +50,10 @@ var (
-	noK8s          = flag.Bool("no-k8s", false, "Run in a non-k8s environment")
+	keysSecretName  = flag.String("keyssecret", "", "Name of the secret to create for generated keys (keys won't be stored unless this is provided)")
+	noK8s           = flag.Bool("no-k8s", false, "Run in a non-k8s environment")
+	metadataTargets = flag.Bool("metadata-targets", true, "Serve individual targets with custom Sigstore metadata")
```
Thanks for adding this flag - Can we add some more wording about this being deprecated in the future in favor of only using the trust root?

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+
+func getTargetUsage(name string) string {
+	for _, knownTargetType := range []string{FulcioTarget, RekorTarget, CTFETarget, TSATarget} {
+		if strings.Contains(name, strings.ToLower(knownTargetType)) {
```
Can we `strings.ToLower(name)` as well?

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+// * CreateRepoOptions.AddMetadataTargets: true
+// * CreateRepoOptions.AddTrustedRoot: false
+func CreateRepo(ctx context.Context, files map[string][]byte) (tuf.LocalStore, string, error) {
+	return CreateRepoWithOptions(ctx, files, CreateRepoOptions{AddMetadataTargets: true, AddTrustedRoot: false})
```
Could we set both to True by default?

bkabrda replied:
My point here was to preserve backwards compatibility of what this function used to do before. If you think that is not an issue, I definitely can do that.

haydentherapper replied:
I'm good with making the change - it adds a file, so it shouldn't be a change that causes compatibility issues.

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+
+	// we sort the targets by Name, this results in intermediary certs being sorted correctly,
+	// as long as there is less than 10, which is ok to assume for the purposes of this code
+	sort.Slice(targets, func(i, j int) bool {
```
I think this is OK, but it does feel a bit fragile since it depends on file naming. Do we have any documentation in this repo we can point to to describe how we expect the names?

haydentherapper replied:
What is setting the file names? I see this is passed in from https://github.com/sigstore/scaffolding/blob/main/cmd/tuf/server/main.go#L94-L126, are file names only controlled by Scaffolding or can they be set by users?

bkabrda replied:
I did this to handle how the [SplitCertChain](https://github.com/sigstore/scaffolding/blob/main/pkg/certs/certs.go#L35) function splits the TSA cert chain.

In terms of who sets the filenames - it's basically a list of files in a directory, so I guess the answer to that is "whoever creates that directory and puts files in there". I don't think we have any documentation about this. Where would such documentation go?

haydentherapper replied:
Could we update https://github.com/sigstore/scaffolding?tab=readme-ov-file#tuf?

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+			hashFunc = crypto.SHA256
+		default:
+			return 0, 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+		}
```
ed25519 as well?

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +150,264 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	return CreateRepoWithOptions(ctx, files, CreateRepoOptions{AddMetadataTargets: true, AddTrustedRoot: false})
+}
+
+func constructTrustedRoot(targets []TargetWithMetadata) (*TargetWithMetadata, error) {
```
Can we add a test for this?

bkabrda replied:
Sure, I will do that.

bkabrda commented:
@haydentherapper I think I addressed all points from your review, except the one about perhaps [the one about some docs](https://github.com/sigstore/scaffolding/pull/1235#discussion_r1728005803) - I can address that if you can advise where to put the docs. Thanks!

haydentherapper commented:
Thanks, I left a comment about where we could put some docs!

bkabrda commented:
@haydentherapper thanks for the pointer, done!

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +151,266 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	now := time.Now()
+
+	// we sort the targets by Name, this results in intermediary certs being sorted correctly,
+	// as long as there is less than 10, which is ok to assume for the purposes of this code
```
Could you add a note about this in documentation?

bkabrda replied:
The README modification that I did says:

```
* `<target>_root.crt.pem`, e.g. `tsa_root.crt.pem`
* `<target>_intermediate_0.crt.pem`, e.g. `tsa_intermediate_0.crt.pem`
* `<target>_intermediate_1.crt.pem`, e.g. `tsa_intermediate_1.crt.pem`
* (more intermediates, but at most 10 intermediate certificates altogether)
* `<target>_leaf.crt.pem`, e.g. `tsa_leaf.crt.pem`
```

I don't see what's missing. Can you be more specific about what I should note?

haydentherapper replied:
I missed the comment about 10 intermediates, all good!

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +151,266 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+		default:
+			return 0, 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+		}
+	case *ed25519.PublicKey:
```
```suggestion
	case ed25519.PublicKey:
```

haydentherapper commented on pkg/repo/repo.go:
```diff
@@ -127,36 +151,266 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	certChain := []*x509.Certificate{}
+
+	// skip potential whitespace at end of file (8 is kinda random, but seems to work fine)
+	for len(rest) > 8 {
```
Could we use https://pkg.go.dev/strings#TrimSpace instead?

bkabrda replied:
Sounds good, I'll do that.

haydentherapper reviewed (commented):
just a few nits

haydentherapper reviewed (dismissed):
just need to rebase

bkabrda commented on pkg/repo/repo.go:
```diff
@@ -127,36 +152,267 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	rest := bytes.TrimSpace(certChainPem)
+	certChain := []*x509.Certificate{}
+
+	// skip potential whitespace at end of file (8 is kinda random, but seems to work fine)
```
Oh, sorry, I forgot about that.

bkabrda commented:
@haydentherapper rebased and addressed the last comment. Thank you for the great review!
